client/hwlib/src/collectors/cpuinfo.rs

```diff
@@ -0,0 +1,183 @@
+/* Copyright 2024 Canonical Ltd.
+ * All rights reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Written by:
+ *        Nadzeya Hutsko <nadzeya.hutsko@canonical.com>
+ */
+
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::{self, BufRead};
+use std::str::FromStr;
+
+use crate::constants::{CPU_MAX_FREQ_FILEPATH, PROC_CPUINFO_FILEPATH};
+
+#[derive(Debug)]
+pub struct CpuInfo {
+    pub platform: String,
+    pub count: usize,
+    pub cpu_type: String,
+    pub model: String,
+    pub model_number: String,
+    pub model_version: String,
+    pub model_revision: String,
+    pub cache: i64,
+    pub bogomips: i64,
+    pub speed: u64,
+    pub other: String,
+}
+
+/// Parse /proc/cpuinfo the same way it's done in checkbox
+/// https://github.com/canonical/checkbox/blob/a8d5e9d/checkbox-support/checkbox_support/parsers/cpuinfo.py
+pub fn parse_cpuinfo(
+    cpuinfo_filepath: Option<&'static str>,
+) -> Result<CpuInfo, Box<dyn std::error::Error>> {
+    let path = cpuinfo_filepath.unwrap_or(PROC_CPUINFO_FILEPATH);
+    let file = File::open(path)?;
+    let reader = io::BufReader::new(file);
+
+    let mut attributes: HashMap<String, String> = HashMap::new();
+    let mut count = 0;
+
+    for line in reader.lines() {
+        let line = line?;
+        if line.trim().is_empty() {
+            continue;
+        }
+
+        let parts: Vec<&str> = line.split(':').collect();
+        if parts.len() == 2 {
+            let key = parts[0].trim().to_string();
+            let value = parts[1].trim().to_string();
+
+            if key == "processor" {
+                count += 1;
+            }
+
+            attributes.insert(key, value);
+        }
+    }
+
+    let machine = std::env::consts::ARCH.to_string();
+    let speed = parse_speed(attributes.get("cpu MHz")).unwrap_or(0);
+
+    let model = attributes
+        .get("Model")
+        .or_else(|| attributes.get("model name"))
+        .unwrap_or(&machine)
+        .to_string();
+
+    let cpu_info = CpuInfo {
+        platform: machine.clone(),
+        count,
+        cpu_type: attributes.get("vendor_id").unwrap_or(&machine).to_string(),
+        model,
+        model_number: attributes
+            .get("cpu family")
+            .unwrap_or(&"".to_string())
```

Comment:
If you [remove](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.remove) the entries you're looking at here then [unwrap_or_default](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_default) should work I think? You're dropping the map at the end of the function anyway so mutating the map as you extract data from it is fine.

Reply:
If I replace `unwrap_or` with `unwrap_or_default`, I get this error

```
the trait `Default` is not implemented for `&std::string::String
```

Am I doing something wrong?

Reply:
you'll need to `remove` the entry rather than using `get`. `get` only returns a reference to the data in the map while `remove` is as it sounds: it removes the data allowing you to own it as a `String` rather than a `&String`.

Reply:
updated in d7d47b8